### PR TITLE
feat: add `/extended/v1/tokens/nft/history` endpoint

### PR DESCRIPTION
**Pull request by rafaelcr**
Adds an endpoint to retrieve the event history of a particular NFT.

The endpoint can:
* Return the `tx_id` or complete `tx` object of the transaction that led that principal to own each NFT
* Return anchored and unanchored results

Examples:

With `tx_id`
```
GET /extended/v1/tokens/nft/history?asset_identifier=SP2X0TZ59D5SZ8ACQ6YMCHHNR2ZN51Z32E2CJ173.the-explorer-guild%3A%3AThe-Explorer-Guild&value=0x01000000000000000000000000000009c5

{
  "limit": 50,
  "offset": 0,
  "total": 2,
  "results": [
    {
      "sender": "SPZGN4KG2AKNNPEWVXXB6EWN1W9WV1JWMGBVVNK3",
      "recipient": "SPNWZ5V2TPWGQGVDR6T7B6RQ4XMGZ4PXTEE0VQ0S.marketplace-v4",
      "event_index": 0,
      "asset_event_type_id": 1,
      "tx_id": "0x4510e6cff6414747f402126b92e35d90165f9f5e1f2750f50ed2b8e052b25108"
    },
    {
      "sender": null,
      "recipient": "SPZGN4KG2AKNNPEWVXXB6EWN1W9WV1JWMGBVVNK3",
      "event_index": 1,
      "asset_event_type_id": 2,
      "tx_id": "0x3ab1ac33ba0a4fb473ee5a2e9d73208a7aceef2d40fea283e3126bda61524ad0"
    }
  ]
}
```

With `tx` metadata
```
GET /extended/v1/tokens/nft/history?asset_identifier=SP2X0TZ59D5SZ8ACQ6YMCHHNR2ZN51Z32E2CJ173.the-explorer-guild%3A%3AThe-Explorer-Guild&value=0x01000000000000000000000000000009c5&tx_metadata=true

{
  "limit": 50,
  "offset": 0,
  "results": [
    {
      "sender": "SPZGN4KG2AKNNPEWVXXB6EWN1W9WV1JWMGBVVNK3",
      "recipient": "SPNWZ5V2TPWGQGVDR6T7B6RQ4XMGZ4PXTEE0VQ0S.marketplace-v4",
      "event_index": 0,
      "asset_event_type_id": 1,
      "tx": {
        "tx_id": "0x4510e6cff6414747f402126b92e35d90165f9f5e1f2750f50ed2b8e052b25108",
        "nonce": 44,
        "fee_rate": "143600",
        "sender_address": "SPZGN4KG2AKNNPEWVXXB6EWN1W9WV1JWMGBVVNK3",
        "sponsored": false,
        "post_condition_mode": "deny",
        "post_conditions": [
          {
            "type": "non_fungible",
            "condition_code": "sent",
            "principal": {
              "type_id": "principal_standard",
              "address": "SPZGN4KG2AKNNPEWVXXB6EWN1W9WV1JWMGBVVNK3"
            },
            "asset": {
              "contract_name": "the-explorer-guild",
              "asset_name": "The-Explorer-Guild",
              "contract_address": "SP2X0TZ59D5SZ8ACQ6YMCHHNR2ZN51Z32E2CJ173"
            },
            "asset_value": {
              "hex": "0x01000000000000000000000000000009c5",
              "repr": "u2501"
            }
          }
        ],
        "anchor_mode": "any",
        "is_unanchored": false,
        "block_hash": "0x236721170b99d5dc0be2e77b5f5e0394ffe49af7862c3e82b91245b50e28d191",
        "parent_block_hash": "0xdf9e6a371ef6706ed9e5362f61981d3e9e1b985acb67495490bacc40391d7372",
        "block_height": 37482,
        "burn_block_time": 1636673415,
        "burn_block_time_iso": "2021-11-11T23:30:15.000Z",
        "parent_burn_block_time": 1636672174,
        "parent_burn_block_time_iso": "2021-11-11T23:09:34.000Z",
        "canonical": true,
        "tx_index": 35,
        "tx_status": "success",
        "tx_result": {
          "hex": "0x0703",
          "repr": "(ok true)"
        },
        "microblock_hash": "",
        "microblock_sequence": 2147483647,
        "microblock_canonical": true,
        "event_count": 1,
        "events": [],
        "execution_cost_read_count": 17,
        "execution_cost_read_length": 50450,
        "execution_cost_runtime": 54533000,
        "execution_cost_write_count": 2,
        "execution_cost_write_length": 685,
        "tx_type": "contract_call",
        "contract_call": {
          "contract_id": "SPNWZ5V2TPWGQGVDR6T7B6RQ4XMGZ4PXTEE0VQ0S.marketplace-v4",
          "function_name": "list-asset",
          "function_signature": "(define-public (list-asset (tradables trait_reference) (tradable-id uint) (price uint) (commission uint)))",
          "function_args": [
            {
              "hex": "0x061605d06be54b4b9fa14cb9bd464635c0bf5287e313127468652d6578706c6f7265722d6775696c64",
              "repr": "SP2X0TZ59D5SZ8ACQ6YMCHHNR2ZN51Z32E2CJ173.the-explorer-guild",
              "name": "tradables",
              "type": "trait_reference"
            },
            {
              "hex": "0x01000000000000000000000000000009c5",
              "repr": "u2501",
              "name": "tradable-id",
              "type": "uint"
            },
            {
              "hex": "0x0100000000000000000000000005f5e100",
              "repr": "u100000000",
              "name": "price",
              "type": "uint"
            },
            {
              "hex": "0x0100000000000000000000000000000064",
              "repr": "u100",
              "name": "commission",
              "type": "uint"
            }
          ]
        }
      }
    },
    {
      "sender": null,
      "recipient": "SPZGN4KG2AKNNPEWVXXB6EWN1W9WV1JWMGBVVNK3",
      "event_index": 1,
      "asset_event_type_id": 2,
      "tx": {
        "tx_id": "0x3ab1ac33ba0a4fb473ee5a2e9d73208a7aceef2d40fea283e3126bda61524ad0",
        "nonce": 37,
        "fee_rate": "81200",
        "sender_address": "SPZGN4KG2AKNNPEWVXXB6EWN1W9WV1JWMGBVVNK3",
        "sponsored": false,
        "post_condition_mode": "deny",
        "post_conditions": [
          {
            "type": "stx",
            "condition_code": "sent_equal_to",
            "amount": "100000000",
            "principal": {
              "type_id": "principal_standard",
              "address": "SPZGN4KG2AKNNPEWVXXB6EWN1W9WV1JWMGBVVNK3"
            }
          }
        ],
        "anchor_mode": "any",
        "is_unanchored": false,
        "block_hash": "0xe90233fb36b6a932de91af1b030723ca71335eed7b0792f2c382a827beb5c2f3",
        "parent_block_hash": "0x3c5657455ed593f16b85619814b10e32040dd13d263f1a423ba04a2c234f8ee9",
        "block_height": 36776,
        "burn_block_time": 1636211530,
        "burn_block_time_iso": "2021-11-06T15:12:10.000Z",
        "parent_burn_block_time": 1636211109,
        "parent_burn_block_time_iso": "2021-11-06T15:05:09.000Z",
        "canonical": true,
        "tx_index": 15,
        "tx_status": "success",
        "tx_result": {
          "hex": "0x0701000000000000000000000000000009c5",
          "repr": "(ok u2501)"
        },
        "microblock_hash": "",
        "microblock_sequence": 2147483647,
        "microblock_canonical": true,
        "event_count": 2,
        "events": [],
        "execution_cost_read_count": 14,
        "execution_cost_read_length": 91190,
        "execution_cost_runtime": 92297000,
        "execution_cost_write_count": 3,
        "execution_cost_write_length": 19,
        "tx_type": "contract_call",
        "contract_call": {
          "contract_id": "SP2X0TZ59D5SZ8ACQ6YMCHHNR2ZN51Z32E2CJ173.the-explorer-guild-mint",
          "function_name": "claim",
          "function_signature": "(define-public (claim ))"
        }
      }
    }
  ]
}
```

Related to #846

### CODE REVIEW

File: docs/entities/tokens/non-fungible-token-history-event-0-tx-id.schema.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    "sender": {
+      "type": "string"
+    },
+    "recipient": {
+      "type": "string"
+    },
+    "event_index": {
+      "type": "integer"
+    },
+    "asset_event_type_id": {
```
Comment (zone117x) [inline docs/entities/tokens/non-fungible-token-history-event-0-tx-id.schema.json:18]: Do we return `asset_event_type_id` and/or its number value elsewhere? I think we generally treat the integer as an internal datastore implementation detail, and we run it through `getAssetEventTypeString` to get a useful string/enum to return on responses.

Comment (rafaelcr) [reply docs/entities/tokens/non-fungible-token-history-event-0-tx-id.schema.json:18]: Nice catch, a string would be much better here, will do

File: docs/entities/tokens/non-fungible-token-history-event-0-tx-id.schema.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "title": "NonFungibleTokenHistoryEventWithTxId",
+  "description": "Non-Fungible Token history event with transaction id",
+  "required": ["event_index", "asset_event_type_id", "tx_id"],
+  "additionalProperties": false,
+  "properties": {
+    "sender": {
+      "type": "string"
```
Comment (zone117x) [inline docs/entities/tokens/non-fungible-token-history-event-0-tx-id.schema.json:10]: What do you think about adding `"nullable": true` for fields like this (like when a token is minted and has no sender)?

It's not technically part of OpenAPI spec, but I think redoc and some other tooling supports it.

Comment (rafaelcr) [reply docs/entities/tokens/non-fungible-token-history-event-0-tx-id.schema.json:10]: That would be cool, I'll look into it

**Review by zone117x — APPROVED** [review]
This looks great @rafaelcr! I didn't spot any obvious bugs, and tests look pretty thorough. 

Left a couple comments/suggestions around the API schema. Other than that, looks good.

**Pull request merged**